sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -3068,6 +3148,17 @@ public WriteResult expand(PCollection<T> input) {
             !getAutoSchemaUpdate(),
             "withAutoSchemaUpdate only supported when using storage-api writes.");
       }
+      if (getRowUpdateFn() != null) {
+        if (method != Method.STORAGE_API_AT_LEAST_ONCE) {
+          checkArgument(
+              getRowUpdateFn() == null,
+              "Row updates only supported when using STORAGE_API_AT_LEAST_ONCE.");
+        }
+        checkArgument(
+            getCreateDisposition() == CreateDisposition.CREATE_NEVER,
+            "CREATE_IF_NEEDED is not supported when applying row updates. Tables must be precreated "
+                + "with a primary key specified.");
+      }
```

Comment:
Does it make sense for non-dynamic destination to fetch the schema and see if the table supports CDC writes?

Reply:
"supports CDC writes" basically means "has a primary key" (today it also requires clustering on those keys, though that requirement is being dropped soon). Unfortunately primary keys are still not exposed via the CRUD API, so we can't check this right now (this is the same reason we don't support CREATE_IF_NEEDED; once there is API support, we could support this creation mode as long as the user specifies a primary key). I believe support in the CRUD api will be there soon, at which point we can do this.

Reply:
The way to do that is to query INFORMATION_SCHEMA (https://cloud.google.com/bigquery/docs/information-schema-table-constraints). The downside is an extra delay, permissions and complexity. Perhaps at some point an optional "withCDCTableValidation" method to make this an explicit request?